libraries/extractor/src/main/java/androidx/media3/extractor/DtsUtil.java

```diff
@@ -223,8 +357,262 @@ public static int getDtsFrameSize(byte[] data) {
+    int headerSizeType = frameBits.readBits(1);
+    int nBits4Header, nBits4ExSSFsize;
+    if (headerSizeType == 0){
+      nBits4Header = 8;
+      nBits4ExSSFsize = 16;
+    }
+    else{
+      nBits4Header = 12;
+      nBits4ExSSFsize = 20;
+    }
+    frameBits.skipBits(nBits4Header); // Extension substream header size
+    int extFrameSize = frameBits.readBits(nBits4ExSSFsize) + 1; // Extension substream frame size
+
+    int numAudioPresent = 1, numAssets = 1;
+    int nRefClockCode = 0, nExSSFrameDurationCode = 0;
+    int staticFieldsPresent = frameBits.readBits(1);
+    if (staticFieldsPresent == 1) {
+      nRefClockCode = frameBits.readBits(2);
+      nExSSFrameDurationCode = 512 * (frameBits.readBits(3) + 1);
+      if (frameBits.readBits(1) == 1) {
+        frameBits.skipBits(32 + 4);
+      }
+
+      int[] nuActiveExSSMask = new int[8];
+      frameBits.skipBits(3 + 3); // numAudioPresent, numAssets
+      for (int nAuPr = 0; nAuPr < numAudioPresent; nAuPr++) {
+        nuActiveExSSMask[nAuPr] = frameBits.readBits(extSSIndex + 1);
+      }
+
+      for (int nAuPr = 0; nAuPr < numAudioPresent; nAuPr++) {
+        for (int nSS = 0; nSS < extSSIndex + 1; nSS++) {
+          if (((nuActiveExSSMask[nAuPr] >> nSS) & 0x1) == 1) {
+            frameBits.skipBits(8); // nuActiveAssetMask
+          }
+        }
+      }
+
+      if (frameBits.readBits(1) == 1) { // bMixMetadataEnbl
+        frameBits.skipBits(2); // nuMixMetadataAdjLevel
+        int bits4MixOutMask = (frameBits.readBits(2) + 1) << 2;
+        int numMixOutConfigs = frameBits.readBits(2) + 1;
+        for (int ns = 0; ns < numMixOutConfigs; ns++) {
+          frameBits.skipBits(bits4MixOutMask); // nuMixOutChMask
+        }
+      }
+    }
+    for (int nAst = 0; nAst < numAssets; nAst++) {
+      int bits4ExSSFsize = (headerSizeType == 0) ? 16 : 20;
+      frameBits.skipBits(bits4ExSSFsize);
+    }
+
+    // Asset descriptor
+    int sampleRate = 0, channelCount = 0;
+    for (int nAst = 0; nAst < numAssets; nAst++) {
+      frameBits.skipBits(9 + 3); // nuAssetDescriptFsize, nuAssetIndex
+      if (staticFieldsPresent == 1) {
+        if (frameBits.readBits(1) == 1) { // bAssetTypeDescrPresent
+          frameBits.skipBits(4); // nuAssetTypeDescriptor
+        }
+        if (frameBits.readBits(1) == 1) { // bLanguageDescrPresent
+          frameBits.skipBits(24); // LanguageDescriptor
+        }
+        if (frameBits.readBits(1) == 1) { // bInfoTextPresent
+          int nuInfoTextByteSize = frameBits.readBits(10) + 1;
+          frameBits.skipBits(nuInfoTextByteSize * 8); // InfoTextString
+        }
+        frameBits.skipBits(5); // nuBitResolution
+
+        sampleRate = SAMPLE_RATE_EXTSS[frameBits.readBits(4)];
+        channelCount = frameBits.readBits(8) + 1;
+      }
+    }
+
+    // Number of audio sample in a compressed DTS frame
+    int sampleCount = nExSSFrameDurationCode * (sampleRate / REF_CLOCK_TABLE[ nRefClockCode]);
```

Comment:
I see the following equation in the spec (page 98):

ExSSFrameDuration = nuExSSFrameDurationCode × RefClockPeriod

Is this line of code we are also multiplying ExSSFrameDuration by the sampleRate? Why is that?

Reply:
The ExSSFrameDuration is the actual duration in seconds. To convert this value to sampleCount(number of samples), we need to multiply it by sampleRate.

Reply:
Because maxSampleRate is:

1. a max value (i.e. not always the same value for every frame and/or every channel)
2. per-channel (p106: "the maximum sample frequency among all channels present in the audio asset.")

Whereas I assume (?) a frame might contain data for multiple(/all?) channels.

So concretely, given:

nuMaxSampleRate = 48_000
nuTotalNumChs = 10
ExSSFrameDuration = 0.05
I would expect (?) a frame could contain 48_000 * 10 * 0.05 = 24_000 samples

Reply:
@rahulnmohan can you please also look at this comment?

Reply:
@rohitjoins 
Sorry, I missed to answer this question.

A frame will contain data for all the channels. If the ExSSFrameDutaion is 0.05, the number of samples decoded per channel will be _Sampling Rate * 0.05_
If the number of channels is 10, then the total sample count will be  _Sampling Rate * 10 * 0.05_

Reply:
@rahulnmohan Agreed that we don't need to multiply this with the number of channels as `ExSSFrameDutaion` is for all channels in the frame. 

Although it is still unclear that the sampling rate we're deriving here is correct, as in the spec it says it is maximum sampling rate (i.e. not always the same value for every frame and/or every channel).

From the spec: `The values that are retrieved from the table are a max sample rate`, How do we justify that?

Reply:
@rohitjoins, the max sampling frequency shall be used as the sample rate of the stream. The Table 'Section G.3.1, G-1: Common attributes' aligns with this point.